Button guidance PR (#4409)
carbon-design-system/carbon-website: src/pages/components/button/usage.mdx

in a variety of ways. Button labels express what action will occur when the user
interacts with it.

> ### When to use

Author: I did a cursory glance on 25% of our components, among the first 10, about 4 components have "When to use" as an H3 and some content below it as "H4".
- [AI label](https://carbondesignsystem.com/components/ai-label/usage/#when-to-use)
- [Checkbox](https://carbondesignsystem.com/components/checkbox/usage/#when-to-use)
- [Content switcher](https://carbondesignsystem.com/components/content-switcher/usage/#when-not-to-use)
- [Dropdown](https://carbondesignsystem.com/components/dropdown/usage/#when-not-to-use)
![image](https://github.com/user-attachments/assets/dabc216c-8273-450b-b941-41b54c50a31e)
![image](https://github.com/user-attachments/assets/5611741a-11ce-4125-babd-9219ba08381f)
![image](https://github.com/user-attachments/assets/9432139a-dbc4-4cdb-882d-ba6a12bac25d)
![image](https://github.com/user-attachments/assets/58b7de6e-db8d-436c-8e04-3fd0eaf0624c)

---

Maybe we can keep a rule that, if there's additional H4 content associated with "When and When not to use" content, use H3, if not then keep it H4.

As per that logic, Button will have it as H4, but screenshots attached above can continue to be the way they are!
Any thoughts on this @aagonzales and @laurenmrice?

> ### When to use

Reviewer: Yeah maybe its that and if you don't have a lot of points to make in a bullet format and its just a sentence or two. I would be fine with keeping the H4 here in your scenario.

Use buttons to communicate actions users can take and to allow users to interact

> with the page. Each page should have only one primary button. Any remaining

Author: I feel that if a user is not going to go to the emphasis section to read more, this brief overview helps right at the top.

</Column>
</Row>

<Caption>Alignment of buttons across various layouts.</Caption>

| Alignment       | Use case                                                                                                                                                        |
| --------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| Left-justified  | Banner call to actions, in-page forms, and nested buttons in components like tiles.                                                                             |
| Right-justified | Inline notifications, inline field buttons and data tables, progressive forms, wizards, and single-button dialogs.                                              |

> | Full-span       | Dialogs, side panel, and small tiles; currently Carbon does not offer a way to implement full-span buttons in code, without an override, they max out at 320px. |

Reviewer 2: Is this part still true? Did we check with a dev? "currently Carbon does not offer a way to implement full-span buttons in code, without an override, they max out at 320px"

```suggestion
| Full-span       | Dialogs, side panel, and small tiles; currently Carbon does not offer a way to implement full-span buttons in code, without an override, they max out at 320px. |
```

#### Fluid button border

> There is [a 1px border](/components/button/style#ghost-button-structure.mdx)

Reviewer 2: Is this link correct? Its going to ghost button structure.

Reviewer 2: Instead of doing the inline link, maybe do the "See the style tab for details" type of thing at the end of the paragraph.

<DoDontRow>
  <DoDont type="do" caption="Do use the {verb} + {noun} content formula in buttons whenever possible.">

![Do use the verb + noun content formula in buttons whenever possible.](images/button_usage_26.png)

  </DoDont>
  <DoDont type="dont" caption="Do not use only a noun as a button label.">

![Do not use only a noun as a button label.](images/button_usage_27.png)

  </DoDont>
</DoDontRow>

<Title>Label alignment</Title>

<DoDontRow>
  <DoDont type="do" caption="Do left-align text in a button, even if the button is wide.">

![Do left-align text in a button, even if the button is wide.](images/button_usage_5.png)

  </DoDont>
  <DoDont type="dont" caption="Do not center align the label in a button, even if the button is wide.">

![Do not center align the label in a button, even if the button is wide.](images/button_usage_6.png)

  </DoDont>
</DoDontRow>

#### Internationalization (RTL)

> For RTL (right-to-left) languages, the entire button is mirrored horizontally.

Reviewer 2: ```suggestion
For right-to-left (RTL) languages, the entire button is mirrored horizontally.
```

  </DoDont>
  <DoDont type="dont" caption="Do not truncate the label of a button if there’s insufficient space inside the button.">

![Do not truncate the label of a button if there’s insufficient space inside the button.](images/button_usage_overflow_2.png)

  </DoDont>
</DoDontRow>

### Further guidance

For further content guidance, see Carbon’s
[content guidelines](https://carbondesignsystem.com/guidelines/content/overview/).

## Universal behaviors

### Primary button focus

The Primary button is the default action. When dialogs appear, the primary
button typically takes focus. On a form, if focus is on a component that is not
actionable with the `Enter` key, pressing `Enter` will activate the primary
button.

<Row>
<Column colLg={8}>

![Primary button takes default focus when there’s no other actionable component in a dialog.](images/button_usage_ub_focus.png)

</Column>
</Row>

<Caption>
  Primary button takes default focus when there’s no other actionable component
  in a dialog.
</Caption>

### States

The default and interactive states of buttons should follow the

> button [Style](/components/button/style#color.mdx) tab guidelines.

Reviewer 2: ```suggestion
button [Style](/components/button/style#color) tab guidelines.
```

### Interactions

#### Mouse

Users can trigger a button by clicking anywhere within the button container.

<Row>
<Column colLg={8}>

![Mouse interaction for button.](images/button_usage_ub_interaction.png)

</Column>
</Row>

<Caption>Mouse interaction for button.</Caption>

#### Keyboard

Users can trigger a button by pressing `Enter` or `Space` while the button has
focus. For additional keyboard interactions, see the
[Accessibility](/components/button/accessibility) tab.

> ### Validation

Reviewer 2: IMO I think we could cut this section since its basically a don't. It could be confusing to user.

tertiary and ghost buttons for these supporting actions or experiences.

<Row>
<Column colLg={8}>

![The primary intent of this page is to present content, with actions like filtering and editing represented by tertiary and ghost buttons.](images/button_usage_primary_3.png)

</Column>
</Row>

<Caption>
  The primary intent of this page is to present content, with actions like
  filtering and editing represented by tertiary and ghost buttons.
</Caption>

## Secondary button

As mentioned in the [Emphasis section](#emphasis),  secondary buttons are
tonally heavy but have less visual prominence than their primary counterparts
due to their lower saturation. As a result, secondary buttons are intended to be
used in conjunction with a primary button. When used in a pair, the secondary
button executes the set's negative action, like "Cancel" or "Back."

<Row>
<Column colLg={12}>

![Secondary button in the context of a product UI.](images/button_usage_secondary_1.png)

</Column>
</Row>

<Caption>Secondary button in the context of a product UI.</Caption>

Secondary buttons should not be used in isolation or for a positive/primary

> action. [Button groups](#button-groups) contain more guidance on how to use
> secondary buttons.

Reviewer 2: Rephrase to fit our standard "See more"  formatting 
```suggestion
action. See [Button groups](#button-groups) for more guidance on how to use
secondary buttons.
```

#### Tertiary buttons for page headers

It is challenging to use a primary button in the header of a page because the
content beneath the header is probably going to have a primary action, or will
in a future release. Even if the the button in the page header is not styled as
a primary button, it has significant prominence due to its hierarchical
placement at the top.. Therefore, it is advised to use a tertiary button for
page headers.

If it is determined that the button in the page header, across all tabs, should
be primary, ensure none of the content below the header contain another primary
action.

<DoDontRow>
  <DoDont type="do" caption="Do use a tertiary button in the page header if there is a primary button in the content section.">

![Do use a tertiary button in the page header if there is a primary button in the content section.](images/button_usage_tertiary_4.png)

  </DoDont>
  <DoDont type="dont" caption="Do not use a primary button in the page header and the content section.">

![Do not use a primary button in the page header and the content section.](images/button_usage_tertiary_5.png)

  </DoDont>
</DoDontRow>

#### Tertiary buttons in button groups

> In [Button groups](#button-groups), where there is one primary and two other

Reviewer 2: I think since this isn't directly reference the "Button group" heading it can just be lower case. 

```suggestion
In [button groups](#button-groups), where there is one primary and two other
```

<Column colLg={8}>

![Danger button in the context of a product UI.](images/button_usage_danger_primary.png)

</Column>
</Row>

<Caption>Danger button in the context of a product UI.</Caption>

<DoDontRow>
  <DoDont type="do" caption="Do use a tertiary danger button for lower emphasis destructive actions.">

![Do use a tertiary danger button for lower emphasis destructive actions.](images/button_usage_danger_tertiary.png)

  </DoDont>
  <DoDont type="do" caption="Do use a ghost danger button for lower emphasis destructive actions.">

![Do use a ghost danger button for lower emphasis destructive actions.](images/button_usage_danger_ghost.png)

  </DoDont>
</DoDontRow>

## Modifiers

### Button with icon

Icons can be placed next to labels to clarify an action and call attention to a
button. However, icons should be used sparingly, as overuse can create visual
noise and make an experience less usable. If you use a button with an icon in
one part of your UI it does not mean that you need to add icons to all other
buttons.

- Use 16px glyphs within buttons; use 20px glyphs within the large expressive
  buttons.
- Glyphs are distinguished by their solid shape and knocked-out details.

> - Glyphs should always appear to the right of the text.

Reviewer 2: Why are we saying "Glyphs" here in all these bullet points, isn't this the rule for icons?

Author: Fixed it with icons now, thanks Anna!

| Copy           | <Copy size={16} aria-label="Copy"/>                | Copy                 |
| Delete         | <TrashCan size={16} aria-label="Trash can"/>       | Trash can            |
| Remove         | <SubtractAlt size={16} aria-label="Subtract alt"/> | Subtract alt         |
| Export         | <Export size={16} aria-label="Export"/>            | Export               |
| Upload         | <Upload size={16} aria-label="Upload"/>            | Upload               |
| Download       | <Download size={16} aria-label="Download"/>        | Download             |
| Play/Start     | <Play size={16} aria-label="Play"/>                | Play                 |
| Pause          | <Pause size={16} aria-label="Pause"/>              | Pause                |
| Stop           | <StopOutline size={16} aria-label="Stop outline"/> | Stop outline         |
| Refresh        | <Restart size={16} aria-label="Restart"/>          | Restart              |

Do you think a universal action with a clearly defined icon is missing? Let us
know [here](https://github.com/carbon-design-system/carbon/issues/new/choose).

> Icons that are not in this list can be used in buttons, as long as the icon

Reviewer 2: I think we should move the sentence up with the other content above the table. These three paragraphs just feel very choppy here. 
<img width="857" alt="image" src="https://github.com/user-attachments/assets/0ec3cea2-32bb-46b1-b208-ce6ea93e6d92" />

Author: Yes, that makes sense! Pushing the second line about icon library above the table. 
Keeping the "let us know here" sentence at the bottom, because it makes sense there!

Let me know if I should move that up as well!

#### Do not use a defined icon to represent a different universal action

Using icons in the above list for other actions can confuse the expected outcome
and experience.

<DoDontRow>
  <DoDont type="do" caption="Do use a defined icon to represent a universal action.">

![Do use a defined icon to represent a universal action.](images/button_usage_button_with_icon_5.png)

  </DoDont>
  <DoDont type="dont" caption="Do not add an icon for an action that would not be associated with it.">

![Do not add an icon for an action that would not be associated with it.](images/button_usage_button_with_icon_6.png)

  </DoDont>
</DoDontRow>

#### Launch icon

The launch icon should be used on any call to action that launches the user into
another tab (whether the content of the new tab is part of the same product or
an entirely separate web resource). Buttons and links requiring the launch icon
are often found in the UI left navigation area, side panel, cards, and modals.
The target destination of the launch action should be made clear to the user
through the button or link label and the surrounding context. For more guidance,

> see [this section](https://pages.github.ibm.com/carbon/ibm-products/guidelines/content/navigation-labels/) 

Reviewer 2: For this link, can we be more specific instead of saying "this section"? It also link to an internal site so we need to put "_For IBMers only:_" at the beginning of the section

of our content guide.

<Row>
<Column colLg={8}>

![Example of a tertiary button with icon indicating that clicking the button would launch a new page.](images/button_usage_button_with_icon_7.png)

</Column>
</Row>

<Caption>
  Example of a tertiary button with icon indicating that clicking the button
  would launch a new page.
</Caption>

> ### Best practices

Reviewer 2: Is this a general "best practices"? Or best practices of icons in buttons? I'm not sure "Best practices should be an H3 under "Modifiers"

  </DoDont>
</DoDontRow>

#### Use the default variation for all icons

The Carbon library contains filled variants of a few icons. Since not every icon
has this variation, we advise using the default option for all icons (the only
exception is status icons, which have their own defined icon to use). Default
icons are named after their action.

<DoDontRow>
  <DoDont type="do" caption="Do use the default icon which is “Play”.">

![Do use the default icon which is “Play”.](images/button_usage_button_with_icon_12.png)

  </DoDont>
  <DoDont type="dont" caption="Do not use the filled icon variations.">

![Do not use the filled icon variations.](images/button_usage_button_with_icon_13.png)

  </DoDont>
</DoDontRow>

### Icon only buttons

Icon only buttons allow users to take actions, and make choices, with a single
tap. Icon buttons can take the form of a primary, secondary, tertiary, or ghost
variant but most commonly will be styled as primary or ghost buttons.

Icon only buttons should be used sparingly. “For most situations, users learn
correct interpretations better with text alone than with icons alone.” —

> Wiedenbeck, S (1999). For this reason, using icon-only buttons is recommended

Reviewer: ```suggestion
Wiedenbeck, S (1999). For this reason, using icon only buttons is recommended
```

> for the following use cases.

Reviewer 2: ```suggestion
for the following use cases:
```

- The icon must be standardized and recognizable without text or must represent
  an action with a strong visual attribute, such as a pin icon for a pinning
  action.
- There is insufficient space and multiple actions, therefore a toolbar using
  icon buttons is required. See
  the [toolbars pattern](https://pages.github.ibm.com/cdai-design/pal/patterns/toolbars)
  for more details.

<Row>
<Column colLg={8}>

![Example of an icon only ghost button in a toolbar showing a tooltip on hover.](images/button_usage_icon_only_button_1.png)

<Caption>
  Example of an icon only ghost button in a toolbar showing a tooltip on hover.
</Caption>

</Column>
</Row>

#### Tooltips for icon only buttons

Regardless of how recognizable an icon may or may not be, or whether that action
lies within the universal actions list, a tooltip is always required with text
explaining what the icon button would do if clicked.

<Row>
<Column colLg={8}>

![Example of an icon only ghost button showing a tooltip on hover.](images/button_usage_icon_only_button_2.png)

<Caption>
  Example of an icon only ghost button showing a tooltip on hover.
</Caption>

</Column>
</Row>

#### Colors of icon only ghost buttons

> Carbon builds type, color and spacing tokens into its components, improving the

Reviewer: ```suggestion
Carbon builds type, color, and spacing tokens into its components, improving the
```
